DataSetExplorer/DataSetBuilder/Model/DataSetProject.cs

```diff
@@ -9,42 +8,60 @@ public class DataSetProject
         public int Id { get; private set; }
         public string Name { get; private set; }
         public string Url { get; private set; }
-        public HashSet<DataSetInstance> Instances { get; private set; }
+        public List<CandidateDataSetInstance> CandidateInstances { get; private set; }
         public DataSetProjectState State { get; private set; }
 
         internal DataSetProject(string name, string url)
         {
             Name = name;
             Url = url;
-            Instances = new HashSet<DataSetInstance>();
+            CandidateInstances = new List<CandidateDataSetInstance>();
             State = DataSetProjectState.Processing;
         }
 
         public DataSetProject(string name) : this(name, null) { }
 
-        internal void AddInstances(List<DataSetInstance> instances)
+        internal void AddCandidateInstance(CandidateDataSetInstance newCandidate)
         {
-            foreach (var instance in instances)
+            var i = CandidateInstances.FindIndex(c => c.CodeSmell.Name.Equals(newCandidate.CodeSmell.Name));
```

Comment:
You can override Equals (and don't forget hashcode) for CandidateInstances to check if the code smell name is the same. Then you can use a HashSet to automatically perform this check (similar to the earlier code).

We can discuss this if you are unsure.